* HCBHA stands for the Hierachical Clustering Based Hybrid Assembly, which could be applied for the `complete` and  `high-quality/high-contiguity` genome reconstruction from  highly complex ecosystems by integrating the iterative strategy (`hybrid assembly, bining and reads mapping`). The genome reconstruction methods require both the short reads and long reads from the same sample. Please note that the iterative strategy is well described in our previous work [`High-Quality Bacterial Genomes of a Partial-Nitritation/Anammox System by an Iterative Hybrid Assembly Method`](https://github.com/Hydro3639/Iterative-Hybrid-Assembly-for-enrichment-system "https://github.com/Hydro3639/Iterative-Hybrid-Assembly-for-enrichment-system")

* For instructions, usage and explanations of the HCBHA approach, please refer to the [`HCBHA wiki`](https://github.com/Hydro3639/HCBHA-STAS/blob/master/HCBHA%20wiki.md "HCBHA wiki")!
* Aside from the evaluation of the proposed iterative HCBHA approach using the Mock dasets, we also demostrated the performance of the approach using a highly complex metagenomic sample, the Activated Sludge sample from Shatin wastewater treatment plant, Hong Kong.
* 
